File: cc9/host/elf2aout.py
#!/usr/bin/env python3
"""
elf2aout — convert a statically-linked x86_64 ELF (laid out by cc9/test/plan9.ld)
into a Plan 9 amd64 a.out.

Plan 9 a.out (amd64): 40-byte big-endian header
  magic(0x8A97) text data bss syms entry spsz pcsz  (8x uint32) + uint64 entry
File: [40B hdr][text][data].  Kernel maps text at TEXTVA=0x200028 and the data
segment at DATAVA = roundup(UTZERO+HDR+textsize, 0x200000); bss is zero-filled
after data. The linker must place text@0x200028 and data@DATAVA so absolute refs
resolve correctly (verified: small-text binary -> data@0x400000).

Usage: elf2aout.py in.elf out.aout
"""
import os, struct, sys

S_MAGIC = 0x8A97
UTZERO  = 0x200000
HDRSZ   = 40
TEXTVA  = UTZERO + HDRSZ          # 0x200028
DATA_ALIGN = 0x200000

def u16(b,o): return struct.unpack_from('<H', b, o)[0]
def u32(b,o): return struct.unpack_from('<I', b, o)[0]
def u64(b,o): return struct.unpack_from('<Q', b, o)[0]

def main():
    elf = open(sys.argv[1], 'rb').read()
    assert elf[:4] == b'\x7fELF' and elf[4] == 2 and elf[5] == 1, 'need LE 64-bit ELF'
    e_entry = u64(elf, 0x18)
    e_phoff = u64(elf, 0x20)
    e_phentsize = u16(elf, 0x36)
    e_phnum = u16(elf, 0x38)

    text_seg = None   # R+X load
    data_seg = None   # R+W load
    for i in range(e_phnum):
        o = e_phoff + i*e_phentsize
        if u32(elf, o) != 1:          # PT_LOAD only
            continue
        flags  = u32(elf, o+0x04)
        seg = dict(off=u64(elf,o+0x08), vaddr=u64(elf,o+0x10),
                   filesz=u64(elf,o+0x20), memsz=u64(elf,o+0x28), flags=flags)
        if flags & 0x1:               # executable -> text
            text_seg = seg
        elif flags & 0x2:             # writable -> data+bss
            data_seg = seg
    assert text_seg, 'no R+X (text) segment'

    # --- text: image based at TEXTVA; linker may bump a few bytes (pad with 0) ---
    tlo = text_seg['vaddr']
    assert tlo >= TEXTVA, f'text vaddr 0x{tlo:x} < 0x{TEXTVA:x}'
    text = bytearray((tlo - TEXTVA) + text_seg['filesz'])
    text[tlo-TEXTVA : tlo-TEXTVA+text_seg['filesz']] = \
        elf[text_seg['off'] : text_seg['off']+text_seg['filesz']]
    text = bytes(text)

    # --- data + bss ---
    if data_seg:
        expect = (TEXTVA + len(text) + DATA_ALIGN - 1) & ~(DATA_ALIGN - 1)
        if data_seg['vaddr'] != expect:
            sys.exit(f"data vaddr 0x{data_seg['vaddr']:x} != kernel-expected "
                     f"0x{expect:x} (text too big for the 0x200000 round?)")
        data = elf[data_seg['off'] : data_seg['off']+data_seg['filesz']]
        bss  = data_seg['memsz'] - data_seg['filesz']
    else:
        data, bss = b'', 0

    # The a.out header packs text/data/bss as 32-bit big-endian. clang (~150 MB)
    # is well under 4 GB, but a segment that overflows would silently truncate and
    # produce a corrupt binary — fail loudly instead.
    for nm, val in (('text', len(text)), ('data', len(data)), ('bss', bss)):
        if val >= (1 << 32):
            sys.exit(f"{nm} segment 0x{val:x} >= 4 GB: does not fit the 32-bit "
                     f"a.out size field (would truncate)")

    # --- Plan 9 amd64 symbol table (so acid can name frames) ---
    # Parse the ELF .symtab/.strtab and emit Plan 9 syms: 8-byte big-endian value,
    # one type byte (0x80|ascii), NUL-terminated name. Text funcs -> 'T', data/bss
    # objects -> 'D'/'B'. Enough for acid backtraces.
    syms = build_symtab(elf, TEXTVA + len(text))

    hdr  = struct.pack('>8I', S_MAGIC, len(text), len(data), bss, len(syms),
                       e_entry & 0xffffffff, 0, 0)
    hdr += struct.pack('>Q', e_entry)
    out = hdr + text + data + syms
    # Atomic write: a plain open('wb') truncates the target FIRST, so a kill mid-write
    # (run9 timeout / OOM) would leave a 0-byte a.out that `deliver`/`run9` then ships
    # (the 0-byte-binary incident). Write a sibling .tmp and os.replace() it — the final
    # path is always either the previous good a.out or the complete new one.
    tmp = sys.argv[2] + '.tmp'
    with open(tmp, 'wb') as f:
        f.write(out)
        f.flush()
        os.fsync(f.fileno())
    os.replace(tmp, sys.argv[2])
    print(f"{sys.argv[2]}: {len(out)} bytes  text={len(text)} data={len(data)} "
          f"bss={bss} syms={len(syms)} entry=0x{e_entry:x}")

def build_symtab(elf, data_end_va):
    e_shoff = u64(elf, 0x28)
    e_shentsize = u16(elf, 0x3a)
    e_shnum = u16(elf, 0x3c)
    symtab = strtab = None
    for i in range(e_shnum):
        o = e_shoff + i*e_shentsize
        if u32(elf, o+4) == 2:            # SHT_SYMTAB
            symtab = (u64(elf, o+24), u64(elf, o+32), u32(elf, o+40))  # off,size,link
    if not symtab:
        return b''
    so, ssz, link = symtab
    lo = e_shoff + link*e_shentsize
    stro, strsz = u64(elf, lo+24), u64(elf, lo+32)
    out = bytearray()
    n = ssz // 24
    for i in range(n):
        e = so + i*24
        st_name = u32(elf, e+0)
        st_info = elf[e+4]
        st_shndx = u16(elf, e+6)
        st_value = u64(elf, e+8)
        typ = st_info & 0xf               # 1=OBJECT 2=FUNC
        if typ not in (1, 2) or st_shndx == 0 or st_value == 0:
            continue
        end = elf.index(b'\x00', stro+st_name)
        name = elf[stro+st_name:end]
        if not name:
            continue
        ch = 'T' if typ == 2 else ('B' if st_value >= data_end_va else 'D')
        out += struct.pack('>Q', st_value) + bytes([0x80 | ord(ch)]) + name + b'\x00'
    return bytes(out)

if __name__ == '__main__':
    main()

File: cc9/host/elf2aout.c
/* elf2aout.c — C port of elf2aout.py, compiled with cc9 to run NATIVELY on 9front
 * (no python9 on the box). Converts a statically-linked x86_64 ELF laid out by
 * cc9/test/plan9.ld into a Plan 9 amd64 a.out. Completes the on-box toolchain
 * (clang -> ld.lld -> elf2aout, all on 9front).
 *
 * Plan 9 a.out (amd64): 40-byte big-endian header
 *   magic(0x8A97) text data bss syms entry32 spsz pcsz (8x u32) + u64 entry
 * File: [40B hdr][text][data]. Kernel maps text@0x200028, data@roundup(...,2MB).
 *
 * Usage: elf2aout in.elf out.aout
 */
extern int  n9_open(const char *, int);
extern int  n9_create(const char *, int, int);
extern long n9_seek(long long *, int, long long, int);
extern long n9_pread(int, void *, long, long long);
extern long n9_pwrite(int, const void *, long, long long);
extern int  n9_close(int);
extern void n9_exits(const char *);
extern void *malloc(unsigned long);

typedef unsigned long u64; typedef unsigned int u32; typedef unsigned short u16;

static void die(const char *m){ n9_pwrite(2, "elf2aout: ", 10, -1); int n=0; while(m[n])n++; n9_pwrite(2,m,n,-1); n9_pwrite(2,"\n",1,-1); n9_exits("elf2aout error"); }
static u16 rd16(unsigned char *b){ return b[0] | (b[1]<<8); }
static u32 rd32(unsigned char *b){ return (u32)b[0] | ((u32)b[1]<<8) | ((u32)b[2]<<16) | ((u32)b[3]<<24); }
static u64 rd64(unsigned char *b){ u64 v=0; for(int i=7;i>=0;i--) v=(v<<8)|b[i]; return v; }
/* big-endian store */
static void be32(unsigned char *p, u32 v){ p[0]=v>>24; p[1]=v>>16; p[2]=v>>8; p[3]=v; }
static void be64(unsigned char *p, u64 v){ for(int i=0;i<8;i++) p[i]=(unsigned char)(v>>((7-i)*8)); }

int main(int argc, char **argv){
	if(argc < 3) die("usage: elf2aout in.elf out.aout");
	int fd = n9_open(argv[1], 0 /*OREAD*/);
	if(fd < 0) die("cannot open input");
	long long sz=0; n9_seek(&sz, fd, 0, 2 /*SEEK_END*/);
	if(sz <= 0) die("empty/seek-fail input");
	unsigned char *elf = (unsigned char *)malloc((u64)sz);
	if(!elf) die("oom reading elf");
	long got = n9_pread(fd, elf, (long)sz, 0); n9_close(fd);
	if(got != (long)sz) die("short read");

	if(!(elf[0]==0x7f && elf[1]=='E' && elf[2]=='L' && elf[3]=='F' && elf[4]==2 && elf[5]==1)) die("need LE 64-bit ELF");
	u64 e_entry = rd64(elf+0x18);
	u64 e_phoff = rd64(elf+0x20);
	u16 e_phentsize = rd16(elf+0x36);
	u16 e_phnum = rd16(elf+0x38);

	const u64 UTZERO=0x200000, HDRSZ=40, TEXTVA=UTZERO+HDRSZ, DATA_ALIGN=0x200000;
	u64 t_off=0,t_vaddr=0,t_filesz=0; int have_t=0;
	u64 d_off=0,d_vaddr=0,d_filesz=0,d_memsz=0; int have_d=0;
	for(u16 i=0;i<e_phnum;i++){
		unsigned char *o = elf + e_phoff + (u64)i*e_phentsize;
		if(rd32(o) != 1) continue;          /* PT_LOAD */
		u32 flags = rd32(o+0x04);
		u64 off=rd64(o+0x08), vaddr=rd64(o+0x10), filesz=rd64(o+0x20), memsz=rd64(o+0x28);
		if(flags & 0x1){ t_off=off; t_vaddr=vaddr; t_filesz=filesz; have_t=1; }
		else if(flags & 0x2){ d_off=off; d_vaddr=vaddr; d_filesz=filesz; d_memsz=memsz; have_d=1; }
	}
	if(!have_t) die("no R+X (text) segment");
	if(t_vaddr < TEXTVA) die("text vaddr < 0x200028");

	u64 tpad = t_vaddr - TEXTVA;            /* leading pad if linker bumped text base */
	u64 textsz = tpad + t_filesz;
	u64 datasz = 0, bss = 0;
	if(have_d){
		u64 expect = (TEXTVA + textsz + DATA_ALIGN - 1) & ~(DATA_ALIGN - 1);
		if(d_vaddr != expect) die("data vaddr != kernel-expected 2MB round");
		datasz = d_filesz; bss = d_memsz - d_filesz;
	}
	if(textsz>=(1ULL<<32) || datasz>=(1ULL<<32) || bss>=(1ULL<<32)) die("segment >= 4GB (32-bit a.out field overflow)");

	/* assemble: [40B hdr][text][data] */
	u64 outsz = HDRSZ + textsz + datasz;
	unsigned char *out = (unsigned char *)malloc(outsz);
	if(!out) die("oom output");
	for(u64 i=0;i<outsz;i++) out[i]=0;
	be32(out+0,  0x8A97);          /* magic */
	be32(out+4,  (u32)textsz);
	be32(out+8,  (u32)datasz);
	be32(out+12, (u32)bss);
	be32(out+16, 0);               /* syms */
	be32(out+20, (u32)(e_entry & 0xffffffff));
	be32(out+24, 0);               /* spsz */
	be32(out+28, 0);               /* pcsz */
	be64(out+32, e_entry);
	/* text (with leading pad already zeroed) */
	for(u64 i=0;i<t_filesz;i++) out[HDRSZ + tpad + i] = elf[t_off + i];
	/* data */
	for(u64 i=0;i<datasz;i++) out[HDRSZ + textsz + i] = elf[d_off + i];

	int ofd = n9_create(argv[2], 1 /*OWRITE*/, 0775);
	if(ofd < 0) die("cannot create output");
	long w = n9_pwrite(ofd, out, (long)outsz, 0); n9_close(ofd);
	if(w != (long)outsz) die("short write");
	n9_pwrite(1, "elf2aout: ok\n", 13, -1);
	n9_exits(0);
	return 0;
}
